scripts/covid_hosp_explore.R

```diff
@@ -137,525 +137,594 @@ forecaster_parameter_combinations_ <- rlang::list2(
+#     list(forecaster = "flatline_fc")
+#   ),
+#   # median forecaster
+#   "ensemble_average",
+#   list(average_type = "median"),
+#   list(
+#     scaled_pop_scaled,
+#     scaled_pop_not_scaled,
+#     smooth_scaled
+#   ),
+#   # mean forecaster with baseline
+#   "ensemble_average",
+#   list(average_type = "mean"),
+#   list(
+#     scaled_pop_not_scaled,
+#     list(forecaster = "flatline_fc")
+#   ),
+#   # median forecaster with baseline
+#   "ensemble_average",
+#   list(average_type = "median"),
+#   list(
+#     scaled_pop_not_scaled,
+#     list(forecaster = "flatline_fc")
+#   )
+# ) %>%
+#   {
+#     if (dummy_mode) {
+#       .$forecasters <- map(.$forecasters, function(x) {
+#         map(x, function(y) {
+#           y$forecaster <- "dummy_forecaster"
+#           y
+#         })
+#       })
+#     }
+#     .
+#   } %>%
+#   mutate(
+#     children_ids = map(.$forecasters, function(x) {
+#       map_chr(x, function(y) {
+#         get_single_id(y)
+#       })
+#     })
+#   ) %>%
+#   add_id(exclude = "forecasters")
+# # spoofing ensembles for right now
+# ensemble_parameter_combinations_ <- tibble::tibble(
+#   id = character(),
+#   ensemble = character(),
+#   ensemble_args = character(),
+#   children_ids = character()
+# )
+# # Check that every ensemble dependent is actually included.
+# missing_forecasters <- setdiff(
+#   ensemble_parameter_combinations_ %>% pull(children_ids) %>% unlist() %>% unique(),
+#   forecaster_grid$id
+# )
+# if (length(missing_forecasters) > 0) {
+#   cli_abort("Ensemble depends on forecasters not included in pipeline: {missing_forecasters}.")
+# }
+# Internals for targets pipeline.
+# Three lists which are used at runtime and must maintain the same order.
+# - forecaster_functions_list: list of forecaster base functions (e.g. scaled_pop)
+# - forecaster_names_list: list of forecaster names (e.g. 'spiniferous.lcont')
+# - params_list: list of parameters for each forecaster (a list of lists)
+# The last one can also be used for parameter lookup, given a forecaster name.
+config$forecaster_functions_list <- config$forecaster_parameter_combinations %>%
+  map(function(x) {
+    x %>% select(forecaster)
+  }) %>%
+  bind_rows() %>%
+  pull(forecaster)
+config$forecaster_names_list <- config$forecaster_parameter_combinations %>%
+  map(function(x) {
+    x %>% select(id)
+  }) %>%
+  bind_rows() %>%
+  pull(id)
+names(config$forecaster_functions_list) <- config$forecaster_names_list
+config$params_list <- config$forecaster_parameter_combinations %>%
+  map(make_params_list) %>%
+  set_names(NULL) %>%
+  unlist(recursive = FALSE)
+names(config$params_list) <- config$forecaster_names_list
+# Create a partially applied forecaster function for each id. This function
+# depends on params_list and forecaster_functions_list, which are defined above.
+# If we don't define these variables here, then targets won't have access at
+# runtime.
+get_partially_applied_forecaster <- function(id) {
```

Comment:
I'm assuming you tried to directly embed this in the `forecast` target? This is messy.

If I'm understanding correctly, effectively what this is doing is embedding the full set of forecasters parameters inside this function, and then using that as the means of passing parameters, instead of passing those in as function arguments to `forecaster_fn`? If that's the case, I worry that literally any changes to any parameters of any forecasters will invalidate all previous runs.

Reply:
I haven't tried embedding directly... seemed like a bit of logic that deserved a wrapper. It's a bit tricky, I should document it more.

But yea, you got it, [this bit of logic was in slide_forecaster before](https://github.com/cmu-delphi/exploration-tooling/pull/169/files/8a9039bf6c52040818d4916a896336adb37c859c#diff-a3b33b551c295811463e722b617cf6d8dd4679aff09d45ecff3e14807f76e3dfL68-L84). The main difference is `forecaster` in that function was already a function object passed from the outside, in this function we use `get()` with a string lookup to find the function from the global environment (e.g. "scaled_pop").